kotlin-formal/KotlinParser.g4

```diff
@@ -0,0 +1,892 @@
+    | nullableType
+    | typeReference
+    | functionType)
+    ;
+
+typeModifiers
+    : typeModifier+
+    ;
+
+typeModifier
+    : annotation | 'suspend' NL*
+    ;
+
+parenthesizedType
+    : '(' NL* type NL* ')'
+    ;
+
+nullableType
+    : (typeReference | parenthesizedType) NL* quest+
+    ;
+
+typeReference
+    : userType
+    | 'dynamic' // do we need a separate dynamic support here?
+    ;
+
+functionType
+    : (receiverType NL* '.' NL*)? functionTypeParameters NL* '->' NL* type
+    ;
+
+receiverType
+    : typeModifiers?
+    ( parenthesizedType
+    | nullableType
+    | typeReference)
+    ;
+
+userType
+    : simpleUserType (NL* '.' NL* simpleUserType)*
+    ;
+
+parenthesizedUserType
+    : '(' NL* userType NL* ')'
+    | '(' NL* parenthesizedUserType NL* ')'
+    ;
+
+simpleUserType
+    : simpleIdentifier (NL* typeArguments)?
+    ;
+
+functionTypeParameters
+    : '(' NL* (parameter | type)? (NL* ',' NL* (parameter | type))* NL* ')'
+    ;
+
+typeConstraints
+    : 'where' NL* typeConstraint (NL* ',' NL* typeConstraint)*
+    ;
+
+typeConstraint
+    : annotation* simpleIdentifier NL* ':' NL* type
+    ;
+
+block
+    : '{' NL* statements NL* '}'
+    ;
+
+statements
+    : (statement (semis statement)* semis?)?
+    ;
+
+statement
+    : (label | annotation)*
+    ( declaration
+    | assignment
+    | loopStatement
+    | expression)
+    ;
+
+declaration
+    : classDeclaration
+    | objectDeclaration
+    | functionDeclaration
+    | propertyDeclaration
+    | typeAlias
+    ;
+
+assignment
+    : directlyAssignableExpression '=' NL* expression
+    | assignableExpression assignmentAndOperator NL* expression
+    ;
+
+expression
+    : disjunction
+    ;
+
+disjunction
+    : conjunction (NL* '||' NL* conjunction)*
+    ;
+
+conjunction
+    : equality (NL* '&&' NL* equality)*
+    ;
+
+equality
+    : comparison (/* NO NL! */ equalityOperator NL* comparison)*
+    ;
+
+comparison
+    : infixOperation (/* NO NL! */ comparisonOperator NL* infixOperation)?
+    ;
+
+infixOperation
+    : elvisExpression (/* NO NL! */ inOperator NL* elvisExpression | isOperator NL* type)*
+    ;
+
+elvisExpression
+    : infixFunctionCall (NL* elvis NL* infixFunctionCall)*
+    ;
+
+infixFunctionCall
+    : rangeExpression (/* NO NL! */ simpleIdentifier NL* rangeExpression)*
+    ;
+
+rangeExpression
+    : additiveExpression (/* NO NL! */ '..' NL* additiveExpression)*
+    ;
+
+additiveExpression
+    : multiplicativeExpression (/* NO NL! */ additiveOperator NL* multiplicativeExpression)*
+    ;
+
+multiplicativeExpression
+    : asExpression (/* NO NL! */ multiplicativeOperator NL* asExpression)*
+    ;
+
+asExpression
+    : prefixUnaryExpression (NL* asOperator NL* type)?
+    ;
+
+prefixUnaryExpression
+    : unaryPrefix* postfixUnaryExpression
+    ;
+
+unaryPrefix
+    : annotation
+    | label
+    | prefixUnaryOperator NL*
+    ;
+
+postfixUnaryExpression
+    : primaryExpression
+    | primaryExpression postfixUnarySuffix+
```

Comment:
Is your intention here to parse "without postfixUnarySuffix first", and only then try parse with `postfixUnarySuffix`?
Unfortunately, current wording causes ANTLR to parse primaryExpression twice.

If that is the case, then single `primaryExpression postfixUnarySuffix+?` probably makes more sense.

If the alternation was carefully picked, then a comment would be really helpful for the readers.

Reply:
IIRC we had some specific reason to do this. 
@shadrina, any ideas?

Reply:
> If that is the case, then single primaryExpression postfixUnarySuffix+? probably makes more sense.

Maybe better `primaryExpression postfixUnarySuffix*`?

Reply:
There is an interesting point about these identical rules. We have an additional thing to consider every time we change something in the grammar — that is, the shape of the tree produced as the result.

You see, the only source of truth we have right now is the kotlin compiler, and the only way to assert that the result of parsing is correct is comparing the shapes of the syntax trees between ANTLR and the compiler. Even small and absolutely correct changes may lead to big changes in the shape of those trees, meaning that even if we don't break anything, we cannot be sure.

Reply:
@KvanTTT , right you are. It might be either `primaryExpression postfixUnarySuffix*?` or `primaryExpression (postfixUnarySuffix+)??`
The latter should be equivalent to the original, however it depends on whatever semantics is preferred.

Reply:
There are some cases where an expression can be parsed both like `primaryExpression` and `primaryExpression` with some suffix. It was an attempt to give a higher priority to the first option. Maybe this can be done in another way, but I remember that `primaryExpression postfixUnarySuffix*` gives unwanted result.

Reply:
Yes, it can be. We faced with [the similar problem](https://github.com/antlr/grammars-v4/blob/master/objc/ObjectiveCParser.g4#L386) in Objective-C grammar. I.e.

```antlr
varDeclaration
    : (declarationSpecifiers initDeclaratorList | declarationSpecifiers) ';'
    ;
```

is not equvalent to

```antlr
varDeclaration
    : declarationSpecifiers initDeclaratorList? ';'
    ;
```

Reply:
@shadrina ,  please note that `(postfixUnarySuffix+)??` is (it should be) different from `postfixUnarySuffix*`.
`postfixUnarySuffix*` is a greedy rule while `(postfixUnarySuffix+)??` is non-greedy rule (i.e. it should try parse without `postfixUnarySuffix+` first, then it would retry with `postfixUnarySuffix+` (which is greedy, so it would try to match as many `postfixUnarySuffix` as it is possible).

Alternative option is `postfixUnarySuffix*?` which is non-greedy, so it would try to parse with NO `postfixUnarySuffix` first, then it would try to parse with one `postfixUnarySuffix` then it would try to parse with two `postfixUnarySuffix`, and so on.

Intuitively, I guess the most reasonable approach should be `postfixUnarySuffix*?`

Reply:
@vlsi, yeap, in that case, `postfixUnarySuffix*?` looks the most appropriate.